### PR TITLE
fix-Control date and time formats from Bika Setup

### PR DESCRIPTION
**Pull request by Nihadness**
First of all, an error due to time formatting appeared on
https://naralabs.atlassian.net/browse/NDEV-22 .

In order to avoid this kind of errors which mostly occurred because of bad-typed inputs from users, it was necessary to define generic Format for Date and Time fields in Bika Setup. It would also let users define their own format independently from Local Time Zone Format of Plone ( which was the actual cause of the issue mentioned above).

Changes have been done to achieve that curation:

- New Records Field was added to Bika Setup, in order to save Formats defined by user.
- AJAX function was added to validate strings from Bika Setup and was used in DateTimeWidget.
- DateTimeField from Plone and its setter method were overriden, for 2 reasons;
     1. For validation.
     2. To make sure DateTime values are set correctly. E.g: For an Instance running on a server which 
        uses any European Time Zone by default, '01/02/2017' could be saved as 2nd of February, where 
        '01/15/2017' could be understood as January 15h. The reason is underlying this, was the 
        constructor of DateTime class. For more information, see: https://pypi.python.org/pypi/DateTime/4.2

- Finally, all DateTimesFields from Plone were replaced with Bika's.

### CODE REVIEW

**Comment by Nihadness** [comment]
@xispa @Espurna , 
Maybe we should close this PR?!
It has 27 files changed, and since we also worked on Sampling Date and Date Sampled fields, obviously there will be too many conflicts.
If this problem is still actual, we can create another branch, cherry-pick possible commits, and then edit rest of the files based on this PR.

**Comment by xispa** [comment]
@Nihadness partially agree. Before taking the decision of closing this PR,  would be nice if you can explain a bit how the caveats described in this PR have been solved because of to the latest contributions in wip. And if there is something that hasn't been solved yet (e.g. "overriding of Plone's DateTime to make sure DateTime values are set correctly, regardless of TimeZone") needs our immediate attention and eventually, a cherry-picking strategy. 

In summary: Does this PR fixes any bug that is still in wip? which one/s?

**Comment by Nihadness** [comment]
@xispa , the main reason of this PR was letting users use Date(time) formats based on their definition in Bika Setup. In some parts of code we have `hardcoded` date formats which must be fixed.
So, I would say, all commits related with `hardcoded` formatting, must be 'cherry-picked'. But it is not easy to merge all of these commits, because of the conflicts (changes) from 'wip'. And if we 'cherry-pick' half of them, and skip some of them,it would be even more confusing for users...
From my point of view, although some commits can be 'cherry-picked', we should not do it directly to 'wip'. We can create a new branch hanging from 'wip', cherry-pick possible commits, finish rest of the work, and then merge it to 'wip'....

Commits which needs an extra work because of changes on 'wip':

- https://github.com/naralabs/bika.lims/pull/176/commits/26a7cb9371a0f9d52776f9976ae25c1156cf3ea9
- https://github.com/naralabs/bika.lims/pull/176/commits/88e791daecd1e6c021a8d3f0795eb233b9b4ce20

Rest of the commits are mostly new code lines, so even they have conflicts it is just because Git cannot merge them.

#Remember
Also need to check if any code line with Date Formatting was added to 'wip' after we started to work on this branch.

**Comment by xispa** [comment]
Ok @Nihadness . Since you are assigned to the ticket `NDEV-22 bika/en-us/date_format_short could not be translated` (which I guess is strongly related with this one), I close this one assuming you will eventually revisit the commits done here when you start to work on NDEV-22

**Comment by xispa** [comment]
@Nihadness please confirm if I can delete this branch. Thanks

**Comment by Nihadness** [comment]
Yes, please do it.
